server/tserver/src/main/java/org/apache/accumulo/tserver/TabletServer.java

```diff
@@ -791,6 +798,57 @@ public void run() {
       }
     }, 0, 5000, TimeUnit.MILLISECONDS);
 
+    // Periodically check that metadata of tablets matches what is held in memory
+    ThreadPools.createGeneralScheduledExecutorService(aconf).scheduleWithFixedDelay(() -> {
+      final SortedMap<KeyExtent,Tablet> onlineTabletsSnapshot = onlineTablets.snapshot();
+
+      final SortedSet<KeyExtent> userTablets = new TreeSet<>();
+      final SortedSet<KeyExtent> nonUserTablets = new TreeSet<>();
+
+      // Create subsets of tablets based on DataLevel: one set who's DataLevel is USER and another
+      // containing the remaining tablets (those who's DataLevel is ROOT or METADATA).
+      // This needs to happen so we can use .readTablets() on the DataLevel.USER tablets in order
+      // to reduce RPCs.
+      // TODO: Push this partitioning, based on DataLevel, to ample.
+      onlineTabletsSnapshot.forEach((k, v) -> {
+        if (Ample.DataLevel.of(k.tableId()) == Ample.DataLevel.USER) {
+          userTablets.add(k);
+        } else {
+          nonUserTablets.add(k);
+        }
+      });
+
+      Map<KeyExtent,Long> updateCounts = new HashMap<>();
+
+      // gather updateCounts for each tablet
+      onlineTabletsSnapshot.forEach((ke, tablet) -> {
+        updateCounts.put(ke, tablet.getUpdateCount());
+      });
+
+      List<TabletMetadata> tmdList;
+
+      // gather metadata for all tablets with DataLevel.USER using readTablets()
+      try (TabletsMetadata tabletsMetadata = getContext().getAmple().readTablets()
+          .forTablets(userTablets).fetch(FILES, LOGS, ECOMP, PREV_ROW).build()) {
+        tmdList = IteratorUtils.toList(tabletsMetadata.iterator());
```

Comment:
Could do this w/ standard java code avoiding a third party dep.

```suggestion
        tmdList = tabletsMetadata.stream().collect(Collectors.toList());
```

Reply:
Good idea. I hadn't considered this.